### ISSUE TITLE
DbContext.ApplyChanges overload accepting lambda

### ISSUE THREAD

**Issue by tonysneed** [TrackableEntities/trackable-entities#106]
We should add an overload to the `DbContext.ApplyChanges` method which accepts a lambda expression, in which the caller could apply logic for setting `EntityState`.  This would allow, for example, the caller to inspect the tracking state or extended tracking state of the entity and then apply some logic for how the `EntityState` should be set.

For example, the caller could determine that, if the [extended tracking state](https://github.com/TrackableEntities/trackable-entities/issues/105) of an entity has been set to `Created` and the entity is being added to an M-M relation, then the `EntityState` should be set to `Added`.  Otherwise, it can be set to `Unchanged`, which will cause it to be added to the M-M relation as an existing entity.

Another possibility would be for the called to check to see if the primary key property of the entity is an identity and the value is zero, and it is in an M-M relation, then the state should be set to `Added` versus `Unchanged`.

This issue was first proposed in #102 and discussed in this [comment](https://github.com/TrackableEntities/trackable-entities/issues/102#issuecomment-170888322).

**Comment by luboshl** [comment]
I started to implement this, but I encountered a problem how to implment that. There is no problem to add overloades for `DbContext.ApplyChanges` with func returning state for given item: `Func<ITrackable, EntityState>`. But what next? If caller implements some logic with some result state, how to use it in the private implementation of `ApplyChanges`? There are so many places where `context.Entry(item).State` is set depending on different conditions...

I created first sample in commit https://github.com/luboshl/trackable-entities/commit/69b586bebe03173bd07ea4fb3dfd7315358504a3

But I think there should also be abillity to set the relation in `context.SetRelationshipState`. The more I think about it the more I mean there must be something more, but I am not sure what...

**Comment by tonysneed** [comment]
I will take a look at this today. Thanks!

**Comment by tonysneed** [comment]
Before I have a look at your sample, let's discuss a general approach, based on your use case of properly applying changes to entities in a M-M relation.  Perhaps we can then write some tests prior to writing an implementation (TDD). :smile:

So let's imagine you are calling `DbContext.ApplyChanges` and passing a callback.  I think the goal of providing a hook to the caller is so that they can then control how TE sets the entity state.  So it seems to me that the callback should return an `EntityState`, so that TE can then blindly set the DbContext entry to that state.  This would provide the caller with ultimate control, which seems like what we want to provide with a hook to `ApplyChanges`, that is, the ability to arbitrarily decide how the entity state should be set, so that the caller could apply some custom or domain-specific logic and then simply tell TE the state that the entity should be set to.

The next question is what information should we supply to the caller so that it can make the decision regarding entity state?  We could start with the M-M scenario.  In order for the caller to determine whether the state should be `Added` or `Unchanged`, we will need to provide the entity itself, so that the caller can either look at `TrackingState` or `TrackingStateExtended`, or it can inspect property values.  We will also need to provide information about the relationship.  So we could add a `RelationshipType` argument, which an enum in TrackableEntities.EF.5/6 that we can make public and put in a separate cs file.

Based on this approach, the callback argument for the overloaded `ApplyChanges` method would be:

```csharp
Func<TEntity, RelationshipType, EntityState>
```

The signature of `ApplyChanges` would be:

```csharp
public static void ApplyChanges<TEntity>(this DbContext context, IEnumerable<ITrackable> items, Func<TEntity, RelationshipType, EntityState> stateSelector)
```

Then in this method we would invoke `stateSelector`, then take the result and either set the entity to the returned `EntityState`, and in the case of an M-M relation, call `context.SetRelationshipState`.

Does this give you a better idea of how to proceed?

If at all possible, it would be great if you could start by writing some failing unit tests.  I will have a look at the tests, then suggest some changes or approve.  Then you can provide a squash commit.  After that, you would work on the implementation.  So please just let me know how you feel about writing some tests, and we can discuss that as well.

**Comment by luboshl** [comment]
The solution is near my proposal (luboshl@69b586b). You added extra argument `RelationshipType` and you made the method `ApplyChanges` generic while I don't and I have the argument `ITrackable` vs. yours `TEntity`. Do you think it is necessary to have it generic? The method iterates through collection of `ITrackable` items so every input in `stateSelector` function always has to be of type `ITrackable`.

`RelationshipType` - I think that type of relationship is known inside `ApplyChanges` method. There is: 
```C#
// If M-M child, set relationship for added or deleted items
if (... && (context.IsRelatedProperty(parent.GetType(), propertyName, RelationshipType.ManyToMany)))
{
  ...
  context.SetRelationshipState(item, parent, propertyName, trackingState.ToEntityState());
  ...
}
```
In `SetChanges` method there is also call to `SetRelationshipState` - for *deleted* items. But again only for M-M relationships.

Maybe I would add `manyToManyRelationshipState` of type `EntityState`. That would be used only for M-M relations and it would explicitly set the state:

```C#
context.SetRelationshipState(item, parent, propertyName, manyToManyRelationshipState);
```

The signature of ApplyChanges would then be:
```C#
public static void ApplyChanges<TEntity>(this DbContext context, IEnumerable<ITrackable> items, Func<TEntity, EntityState, EntityState> stateSelector)
```
What do you think about this?

**Tests** - I am not sure its possible to create failing test, because this is not a bug. I could prepare a test with scenario where new items in both sides of M-M reference are added and this test fails. But when we implement this new feature, test would still fail and it would need to reimplement with this new feature (overloaded `ApplyChanges`). Or did you meant something else?

**Comment by tonysneed** [comment]
Good discussion, here are my responses ...

> Do you think it is necessary to have it generic? The method iterates through collection of ITrackable items so every input in stateSelector function always has to be of type ITrackable.

Yes, it needs to be generic because the caller of the method needs to be able to use the entity in a type-safe manner.  For example, if I am calling `ApplyChanges` on a `Product`, I may wish to look at the `ProductId` property to see if it is zero and, if so, return an `EntityState` of Added.  If we just pass an `ITrackable` the caller would have to cast it to `Product`, which is not type-safe.  And the entity will implement `ITrackable` or `ITrackableExtended`, so `TrackingState` would be available without needing to cast.

> RelationshipType - I think that type of relationship is known inside ApplyChanges method.

Yes, so we need to create a separate RelationshipType.cs file, put the `RelationshipType` enum in there and make it public.

> Maybe I would add manyToManyRelationshipState of type EntityState.

So a clarification is that `EntityState` is not a TE type.  It belongs to Entity Framework as [System.Data.EntityState])https://msdn.microsoft.com/en-us/library/system.data.entitystate(v=vs.110).aspx).  We are just using it, just as we're using other EF types, such as DbContext, etc.

> Tests - I am not sure its possible to create failing test, because this is not a bug.

We're taking a [TDD approach](https://msdn.microsoft.com/en-us/library/aa730844(v=vs.80).aspx) with this issue, so what that means is:

1. We can first agree on the method signature for the `ApplyChanges` overload.
2. After making sure you have pulled latest for the *develop* branch, create a new local branch from *develop*.
3. Add the `ApplyChanges` method overload using the agreed-upon signature, and make 
4. Then we write some tests for the new `ApplyChanges` method, which `RelationshipType` public in a separate file.
5. Commit these changes, publish your branch, then submit a new PR.
6. Add tests which call the new `ApplyChanges` method, and which assert that we are getting the expected result for how we arranged things.  For example, in the callback we assert that the EntityState has been set to `Added` for an entity when we return `Added` from the callback.  These tests will *fail*, because no implementation has been written for the new `ApplyChanges` method.
7. Commit the addition of the tests, then push your commit.
8. I will get your commits and give feedback.  You can make any requested changes, commit them, then push.  We'll do this until we have a nice set of failing tests and we think we have the main scenarios covered.
9. Once we are satisfied with the specs, you can then proceed to write the implementation, which we can discuss on the PR.  You can make as many commits as you want in this process.  The goal will be to get all the tests to pass.
10. I'll merge the commit, and we're done!

**Comment by luboshl** [comment]
**Generic method** - you are right, this is better.

`RelationshipType` - I mean sometehing else. When an item is processed in `ApplyCanges` method, it is known, which type of relationship is current item in. Whether it is 1-1, 1-M etc. We only want to be able to set explicitly state of the `relationship` when it is M-M. We are not interested in other relationships. And in this case we need to set the state of the relationship with this method call: 
```C#
context.SetRelationshipState(item, parent, propertyName, manyToManyRelationshipState);
```
Notice the last argument `manyToManyRelationshipState` from my previous comment (https://github.com/TrackableEntities/trackable-entities/issues/106#issuecomment-172345587). I understand, that `EntityState` is type from EF. It is just the type that is expected in `context.SetRelationshipState()` extension method as the last argument.

I don't understand, what to do with argument of enum type `RelationshipType`. I think we don't need it here. It says what kind of relationship it is, but it says nothnig about final state in relationship. Are we clear?

All 10 steps are understandable for me. (Except 3. - *"and make"* what? :smile:) Now it makes sense to me to prepare failing test - it will be done **after** new overload of `ApplyChanges`, not **before** it. (But before implementaion itself.)

**Comment by luboshl** [comment]
Ups, I thought about the `stateSelector` as it has 2 outputs!!! It's nonsense, of course... :rage:

My ideas was - for an item (entity) set entity state and relation state. To accomplish that, I need two callbacks.

And in your proposal - you mean to have entity **and** `RelationshipType` as an input for the function?

**Comment by tonysneed** [comment]
Let me walk through the sequence of calls, which I think may make things clearer. :ok_hand: 

Assuming we have the method signature as follows:

```csharp
public static void ApplyChanges<TEntity>(this DbContext context, ITrackable item, Func<TEntity, RelationshipType, EntityState?> stateSelector)
```

The caller would invoke `ApplyChanges` as follows:

```csharp
context.ApplyChanges<Product>(product, (e, rt) => {
    if (rt == RelationshipType.ManyToMany)
    {
        if (e.ProductId == 0) {
            return EntityState.Added; }
        return EntityState.Unchanged;
    }
    return null;
}
```

You can see here that `ApplyChanges` will call back the caller via the lambda expression, passing two parameters to the caller: `e` which is a `TEntity` and `rt` which is a `RelationshipType`.  The caller then uses these two pieces of information to decide what the `EntityState` of the entity should be, and this `EntityState` is then returned to `ApplyChanges` where the entity state is set accordingly.

As you can see here, the goal is to make `ApplyChanges` general-purpose, so that the caller can have a way to instruct TE how to set the entity state.  This is using what is called the [Strategy Pattern](https://en.wikipedia.org/wiki/Strategy_pattern).  We don't want to limit it to just M-M relations, but we want to call the caller to tell TE how to set entity state with any kind of relation.

Does this help clarify the design and the flow of information between the caller and `ApplyChanges`? :pray: 

We haven't yet discussed the code in `ApplyChanges`, but that's OK, because we are using TDD.  At this point all we really care about is the method signature and how it is to be used.  Once we get past the design, we can then talk about the code in `ApplyChanges` which would invoke the callback and then process the result.

**Comment by luboshl** [comment]
So what about this two options:

#### 1) One state selector
```C#
public static void ApplyChanges<TEntity>(this DbContext context,
    IEnumerable<ITrackable> items,
    Func<TEntity, ExplicitStatesHolder> stateSelector)
```
... where `ExplicitStatesHolder` is this class:
```C#
public class ExplicitStatesHolder
{
    public EntityState? EntityState { get; set; }
    public EntityState? ManyToManyRelationshipState { get; set; }

    public ExplicitStatesHolder(EntityState? entityState, EntityState? manyToManyRelationshipState)
    {
        this.EntityState = entityState;
        this.ManyToManyRelationshipState = manyToManyRelationshipState;
    }
}
```

#### 2) Two state selectors
```C#
public static void ApplyChanges<TEntity>(this DbContext context,
    IEnumerable<ITrackable> items,
    Func<TEntity, EntityState> entityStateSelector,
    Func<TEntity, EntityState> manyToManyRelationshipStateSelector)
```

Two notes:
* "ExplicitStatesHolder" is silly name. Naming is always problem for me :smile: 
* I forgot second note... :smile: 

I personally prefer first option with only one state selector.

**Comment by tonysneed** [comment]
Are you responding to my latest comment? I think our comments may have crossed paths. :innocent:

`ApplyChanges` need to send the relationship type to the caller, so that the caller can decide what `EntityState` to return to `ApplyChanges`.  The caller only needs to return a single `EntityState` to `ApplyChanges` because that's the state which `ApplyChanges` should set the state to.

Does my example above help?

**Comment by luboshl** [comment]
To answer your comment https://github.com/TrackableEntities/trackable-entities/issues/106#issuecomment-172376855 - usage of `RelationshipType` is absolutely clear now. I understood the ussage of the method `ApplyChanges` with lambda. What I didn't know, that we want to use it for all relationships, not only for M-M :smile: But of course, it makes sense. So it answers my last question in previous comment:

> And in your proposal - you mean to have entity **and** `RelationshipType` as an input for the function?

**Comment by luboshl** [comment]
>Are you responding to my latest comment? I think our comments may have crossed paths.

No, it was realy comment to 'before last' comment :smile:

**Comment by tonysneed** [comment]
Great, it sounds like we're all clear then? :pray: 

Yes, the Func callback would pass both the entity and the RelationshipType, so the caller can use them to decide what `EntityState` to return.  Notice that `EntityState` is nullable, because returning `null` would indicate to `ApplyChanges` that it should keep the default behavior.

**Comment by luboshl** [comment]
Ok, all clear :+1:

**Comment by tonysneed** [comment]
And I corrected [3] in my [comment](https://github.com/TrackableEntities/trackable-entities/issues/106#issuecomment-172360289).

I'm looking forward to getting your first unit test!  You can put tests in another region in the `NorthwindDbContextTests` class in TrackableEntities.EF.5.Tests.

**Comment by luboshl** [comment]
>I'm looking forward to getting your first unit test! You can put tests in another region in the NorthwindDbContextTests class in TrackableEntities.EF.5.Tests.

So here you are: https://github.com/TrackableEntities/trackable-entities/commit/52cce30cba70014f9a0a35e7ebea4f13ba0edd5b :satisfied:

And PL https://github.com/TrackableEntities/trackable-entities/pull/111 is also done.

**Comment by luboshl** [comment]
I added two other (failing) tests - this time for **one to many** relationship. What do you think about this usage?
```C#
var context = TestsHelper.CreateNorthwindDbContext(CreateNorthwindDbOptions);
var order = new MockNorthwind().Orders[0];
var detail1 = order.OrderDetails[0];
var detail2 = order.OrderDetails[1];
var detail3 = order.OrderDetails[2];

var id1 = detail1.OrderDetailId;
var id2 = detail2.OrderDetailId;
var id3 = detail3.OrderDetailId;

context.ApplyChanges<ITrackable>(order, (e, rt) =>
{
    var orderEntity = e as Order;
    var orderDetailEntity = e as OrderDetail;

    if (orderEntity != null)
        return EntityState.Added;

    if (orderDetailEntity != null)
    {
        if (orderDetailEntity.OrderDetailId == id1)
            return EntityState.Added;

        if (orderDetailEntity.OrderDetailId == id2)
            return EntityState.Modified;

        if (orderDetailEntity.OrderDetailId == id3)
            return EntityState.Deleted;
    }

    return null;
});
```

I want to check two different types of entity, `Order` and `OrderDetail`, which are processed in one call to `AcceptChanges`. So I cannot use (implicit or explicit) generic argument `<Order>`, because I need to set state also for `OrderDetail`. So I set `<ITrackable>` as a generic argument. Then, unfortunately, I have to cast argument `e`.

**Comment by tonysneed** [comment]
Now that we have PR #111, let's continue the discussion over there, where we can link back to comments in this issue.

**Comment by tonysneed** [comment]
Resolved with the merging of PR #118.
